### ISSUE TITLE
Full framework test csproj referencing netstandard csproj using HttpClient fails at runtime

### ISSUE THREAD

**Issue by jnm2** [dotnet/corefx#13619]
I have a net461 test project that uses a netstandard 1.3 csproj. The netstandard project uses HttpClient.

When I run the tests, some with HTTP gets and some with HTTPS, they fail with this:

```
System.IO.FileLoadException : Could not load file or assembly 'System.Security.Cryptography.X509Certificates, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
   at System.Net.Http.WinHttpHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClientHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpMessageInvoker.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.SendAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.GetAsync(Uri requestUri, HttpCompletionOption completionOption, CancellationToken cancellationToken)
```

What can I do about this? There are no warnings or other hints to point me in any direction, and Google searches have left me more confused than ever.

Windows 10 x64 AU
Microsoft Visual Studio Professional 2015
Version 14.0.25431.01 Update 3
Microsoft .NET Framework
Version 4.6.01586
Microsoft .NET Core Tools (Preview 2)   14.1.20907.0

Don't know if this matters, dotnet:
  Version  : 1.0.1
  Build    : cee57bf6c981237d80aa1631cfe83cb9ba329f12

**Comment by karelz** [comment]
Looks similar to #13409 - I think you need new NuGet.
cc: @ericstj

**Comment by jnm2** [comment]
If that's the case, we seriously need a system that gives package restore errors if you're on an unsupported version of NuGet.

**Comment by jnm2** [comment]
I installed the NuGet 3.5 vsix and restarted VS, restored packages, and ran tests. Same result.

**Comment by ericstj** [comment]
> Looks similar to #13409 - I think you need new NuGet.

No, not the same.

> net461 test project that uses a netstandard 1.3 csproj. The netstandard project uses HttpClient

What's happening here is you have some dependencies that exclusive to a project reference, and those aren't being applied to the referencing project.

Transitive flow of dependencies is a new thing with NuGet 3 and still rather fragile.  That's one of the things that's NuGet is working on making better with native `<PackageReference>` support in MSBuild.

What does your project reference look like?  Is it a normal `<ProjectReference>` between two CSProj's?  Are both CSProj's using project.json?

I suspect the `net461` project is using packages.config.  If that's true then you have to install all NuGet packages manually.  NuGet-packages.config doesn't flow packages across project references.  Whenever working with packages.config you always have to be careful to install all packages in the referencing project that are installed in the referenced project.

As I mentioned project.json added this support so converting the `net461` project to use `project.json` is another option.

Let me know if this helps.

**Comment by jnm2** [comment]
> What does your project reference look like? Is it a normal <ProjectReference> between two CSProj's? Are both CSProj's using project.json?

Yes and yes. I use project.json for all csprojs. Should I upload a skeleton example?

**Comment by ericstj** [comment]
What are you using to restore?  NuGet/VS or dotnet.exe?  (the latter doesn't understand p2p references between CSProjs).

Sure, if you have an example that'd help.

**Comment by jnm2** [comment]
NuGet/VS. 

https://github.com/jnm2/misc-codesamples/raw/master/Bug%20reports/.NET/Netstandard%20HttpClient%20issue.zip

To reproduce, open the solution and run all tests in the Visual Studio test window.

**Comment by ericstj** [comment]
Oops, I missed the part of error, which @karelz did not: `The located assembly's manifest definition does not match the assembly reference`

That's because the test is running without bindingRedirects.  I made the following changes to the test CSproj to fix this: 
```xml
    <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
    <GenerateBindingRedirectsOutputType>true</GenerateBindingRedirectsOutputType>
```

Related issue on MSBuild is here: https://github.com/Microsoft/msbuild/issues/1310

Can you try this and let me know if it fixes it for you?

**Comment by jnm2** [comment]
The test appears to run indefinitely, and when I debug I get this:

> System.IO.FileLoadException occurred
> Message: Exception thrown: 'System.IO.FileLoadException' in System.Net.Http.dll
Additional information: Could not load file or assembly 'Microsoft.Win32.Primitives, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)

```
	System.Net.Http.dll!System.Net.Http.WinHttpHandler.StartRequest(object obj)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncVoidMethodBuilder.Start<System.Net.Http.WinHttpHandler.<StartRequest>d__101>(ref System.Net.Http.WinHttpHandler.<StartRequest>d__101 stateMachine)	Unknown
 	System.Net.Http.dll!System.Net.Http.WinHttpHandler.StartRequest(object obj)	Unknown
 	System.Net.Http.dll!System.Net.Http.WinHttpHandler.SendAsync.AnonymousMethod__97_0(object s)	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.InnerInvoke()	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.Execute()	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.ExecutionContextCallback(object obj)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.ExecuteWithThreadLocal(ref System.Threading.Tasks.Task currentTaskSlot)	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.ExecuteEntry(bool bPreventDoubleExecution)	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()	Unknown
 	mscorlib.dll!System.Threading.ThreadPoolWorkQueue.Dispatch()	Unknown
 	mscorlib.dll!System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()	Unknown
 	[Native to Managed Transition]	
```

I tried the two properties first in the top property group and then in a new property group last thing in the file.

**Comment by ericstj** [comment]
It looks like MSBuild is missing the conflict and not generating a redirect.  I haven't gotten to the bottom of why that is happening.

**Comment by ericstj** [comment]
I found what's wrong.  MSBuild misses the conflict because System.Net.Http.dll's reference assembly is does not reference to Microsoft.Win32.Primitives.  This is related to https://github.com/dotnet/corefx/issues/7702 but the workaround there won't fix this.  You could add a bindingRedirect directly in the test project as a workaround, but ultimately I think MSBuild needs to improve this.  It should be examining runtime assemblies as well for bindingRedirects, not just references.

**Comment by jnm2** [comment]
Something that can be fixed in time for 1.1?

**Comment by ericstj** [comment]
It's a problem with MSBuild, not .NET Core.  We could use the same workaround strategy I mentioned in https://github.com/dotnet/corefx/issues/7702 for 1.1, but can't really fix the root problem.

This class of problem will be fixed in general with the NETStandard changes that are coming.  We'll be coalescing all these to a single assembly (netstandard.dll) which will be garunteed to be in the compile closure, so MSBuild will always see a conflict.

**Comment by jnm2** [comment]
I think I'm following- so it will never be possible to fix msbuild so that this scenario with netstandard 1.3 just works?

**Comment by ericstj** [comment]
It's possible, but requires updates to MSBuild, so it would be a release like Visual Studio to fix those.  I have filed some issues in the MSBuild repo that could help here.

Meanwhile we could create a package that works with netstandard1.3 and makes it "just work" with the old MSBuild by forcing it to see conflicts, even if you aren't using the conflicting assemblies at compile time.

**Comment by jnm2** [comment]
>  I have filed some issues in the MSBuild repo that could help here.

I'd like to follow them. Just this one? https://github.com/Microsoft/msbuild/issues/1344

As long as sometime in the future netstandard1.3 just works like I was expecting, I'm willing to implement a workaround.

**Comment by ericstj** [comment]
That one and this https://github.com/Microsoft/msbuild/issues/1310

**Comment by jnm2** [comment]
What's my best bet for the time being?

**Comment by karelz** [comment]
If you're asking when msbuild will fix the problem - it would be best to ask them :)
Are we ok to close this question as answered?

**Comment by jnm2** [comment]
I'm actually asking if a workaround exists or if I should skip testing net461?

**Comment by ericstj** [comment]
You can workaround by adding the bindingRedirect for Microsoft.Win32.Primitives to an app.config in the test project.  0.0.0.0-4.0.1.0 -> 4.0.1.0.  I'm leaving https://github.com/dotnet/corefx/issues/7702 open to track a better workaround.

**Comment by jnm2** [comment]
Thank you, I'm unblocked.

**Comment by Trolldemorted** [comment]
I think i am currently experiencing this issue.

One of our .NETStandard1.5 libraries depends on `System.Net.Http`, which according to NuGet depends on `System.Security.Cryptography.X509Certificates (>= 4.1.0)` on `.NETFramework4.6`.
The testing project (.NETFramework4.6.2) references `System.Security.Cryptography.X509Certificates v4.3.0`.
When the library runs into X509-code, it throws the same exception jnm2 has posted (for version 4.0.0.0).

@ericstj how do i add a correct bindingRedirect?
i added this to my app.config, restarted VS, rebuilt the solution, but it still throws the same exception:
```
<dependentAssembly>
  <assemblyIdentity name="Microsoft.Win32.Primitives"/>
    <bindingRedirect oldVersion="0.0.0.0-4.0.1.0" newVersion="4.0.1.0" />
  </dependentAssembly>
</assemblyBinding>
```

**Comment by ericstj** [comment]
@Trolldemorted did you also make the project changes mentioned above?
https://github.com/dotnet/corefx/issues/13619#issuecomment-260425946

**Comment by Trolldemorted** [comment]
No, didn't know that was also required, sorry.

I have put the elements into the first unconditional PropertyGroup and now the project won't build any more:
```
2>------ Neues Erstellen gestartet: Projekt: libsignal-service-dotnet-tests, Konfiguration: Debug Any CPU ------
2>C:\Program Files (x86)\MSBuild\14.0\bin\Microsoft.Common.CurrentVersion.targets(1898,5): error MSB4018: Unerwarteter Fehler bei der GenerateBindingRedirects-Aufgabe.
2>C:\Program Files (x86)\MSBuild\14.0\bin\Microsoft.Common.CurrentVersion.targets(1898,5): error MSB4018: System.NullReferenceException: Der Objektverweis wurde nicht auf eine Objektinstanz festgelegt.
2>C:\Program Files (x86)\MSBuild\14.0\bin\Microsoft.Common.CurrentVersion.targets(1898,5): error MSB4018:    bei Microsoft.Build.Tasks.GenerateBindingRedirects.UpdateExistingBindingRedirects(XElement runtimeNode, IDictionary`2 redirects)
2>C:\Program Files (x86)\MSBuild\14.0\bin\Microsoft.Common.CurrentVersion.targets(1898,5): error MSB4018:    bei Microsoft.Build.Tasks.GenerateBindingRedirects.Execute()
2>C:\Program Files (x86)\MSBuild\14.0\bin\Microsoft.Common.CurrentVersion.targets(1898,5): error MSB4018:    bei Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
2>C:\Program Files (x86)\MSBuild\14.0\bin\Microsoft.Common.CurrentVersion.targets(1898,5): error MSB4018:    bei Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()
```

**Comment by ericstj** [comment]
That appears to be a bug with [GenerateBindingRedirects](https://github.com/Microsoft/msbuild/blob/cb8c727a898e4439a3cd18f329504e436b0b7c00/src/XMakeTasks/AssemblyDependency/GenerateBindingRedirects.cs#L186).  File an issue on MSBuild for that.  They should never throw a NRE.

I think you're missing some attributes on the `assemblyIdentity` element.  Have a look and see if you can spot it.  If not share your app.config somewhere (eg: as a gist).

**Comment by Trolldemorted** [comment]
https://gist.github.com/Trolldemorted/8a2d467d1c46499717b9eb4eb27cbd8a

There you go, app.config and the .csproj.

Will the MSBuild team require anything except for these two files? I am thinking about deleting the test project and creating a new one in case i did something horribly wrong with NuGet.

Are earlier versions of .NETFramework also affected by this? I could switch to any version that supports .NETStandard1.4, i think.

**Comment by ericstj** [comment]
```
<assemblyIdentity name="Microsoft.Win32.Primitives"/>
```
should be
```
<assemblyIdentity name="Microsoft.Win32.Primitives" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
```

**Comment by Trolldemorted** [comment]
Thanks for your patience, now the tests are running! How would i obtain the token if i have to do this again?

**Comment by ericstj** [comment]
@Trolldemorted no worries, glad it was something simple (ish).  We're aiming to improve on a lot of this in the upcoming releases.

`sn -t <pathToAssembly>` will give you the public key token.  sn.exe is in a developer command prompt.

You can also right click the reference in VS and view properties and you'll see the full strong name.

**Comment by Trolldemorted** [comment]
Now i included said library into a Xamarin project (which i converted to .NETStandard according to [this guide](https://xamarinhelp.com/dot-net-standard-pcl-xamarin-forms/)), compiling works fine but when i run the UWP app the same error happens for `System.Runtime, Version=4.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a`.

The Xamarin project(s) don't have an app.config - what should i do in this case?

**Comment by emirhosseini** [comment]
I'm having the same issue with the Visual Studio 2017 test explorer.  Running a "dotnet test" command however works fine.  Running on CLI version rc3.
